### PR TITLE
Fix: set the default CMake generator and toolset for Intel C++

### PR DESCRIPTION
**Pull request by SSE4**
/cc @ohanar
related: #5699
based on https://github.com/ohanar/conan/commit/7a62f3ac35573fdc6aa6a25d44111b87c726cb90
tested locally with the following conan profile:
```
[settings]
arch=x86_64
arch_build=x86_64
os=Windows
os_build=Windows
compiler=intel
compiler.version=19.1
compiler.base=Visual Studio
compiler.base.version=16
compiler.base.runtime=MDd
build_type=Debug
```
(Intel C++ 19.1)

as result, conan sets correct CMake generator and toolset, and CMake successfully uses the right compiler:
```
log4cplus/2.0.4 (test package): Calling build()
-- The C compiler identification is Intel 19.1.0.20200306
-- The CXX compiler identification is Intel 19.1.0.20200306
-- Check for working C compiler: C:/Program Files (x86)/IntelSWTools/compilers_and_libraries_2020/windows/bin/intel64/icl.exe
-- Check for working C compiler: C:/Program Files (x86)/IntelSWTools/compilers_and_libraries_2020/windows/bin/intel64/icl.exe -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: C:/Program Files (x86)/IntelSWTools/compilers_and_libraries_2020/windows/bin/intel64/icl.exe
-- Check for working CXX compiler: C:/Program Files (x86)/IntelSWTools/compilers_and_libraries_2020/windows/bin/intel64/icl.exe -- works
```

Changelog: Fix: Set the default CMake generator and toolset for Intel C++.
Docs: omit

- [x] Refer to the issue that supports this Pull Request.
- [x] If the issue has missing info, explain the purpose/use case/pain/need that covers this Pull Request.
- [x] I've read the [Contributing guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).
- [x] I've followed the PEP8 style guides for Python code.
- [ ] I've opened another PR in the Conan docs repo to the ``develop`` branch, documenting this one. 

<sup>**Note:** By default this PR will skip the slower tests and will use a limited set of python versions. Check [here](https://github.com/conan-io/conan/blob/develop/.github/PR_INCREASE_TESTING.md) how to increase the testing level by writing some tags in the current PR body text.</sup>

### CODE REVIEW

File: conans/client/build/cmake_flags.py [hunk 1]
```diff
@@ -19,10 +19,15 @@
...
-    if settings.get_safe("compiler") == "Visual Studio":
+    compiler = settings.get_safe("compiler")
+    compiler_base = settings.get_safe("compiler.base")
+    if compiler == "Visual Studio":
         subs_toolset = settings.get_safe("compiler.toolset")
         if subs_toolset:
             return subs_toolset
+    elif compiler == "intel" and compiler_base == "Visual Studio":
+        compiler_version = settings.get_safe("compiler.version")
+        return "Intel C++ Compiler " + compiler_version
```
Comment (ohanar) [inline conans/client/build/cmake_flags.py]: Need this to be (e.g.) "Intel C++ Compiler 19.0" not "Intel C++ Compiler 19". Also, if "compiler.version" is not set, this function should probably return `None`.

Comment (ohanar) [reply conans/client/build/cmake_flags.py]: This change also breaks when an explicit generator is set that is not Visual Studio project files. I think the issue the root cause is [here](https://github.com/SSE4/conan/blob/intel_toolset/conans/client/build/cmake.py#L177), which should add a `"Visual" in self.generator` check.

**Review by ohanar — CHANGES_REQUESTED** [review]

File: conans/client/build/cmake_flags.py (hunk 1 above)
Comment (SSE4) [reply conans/client/build/cmake_flags.py]: addressed both problems, added an additional test, thanks for the catching

File: conans/client/build/cmake_flags.py [hunk 2]
```diff
@@ -18,11 +18,19 @@
...
+    if compiler == "Visual Studio":
         subs_toolset = settings.get_safe("compiler.toolset")
         if subs_toolset:
             return subs_toolset
+    elif compiler == "intel" and compiler_base == "Visual Studio" and "Visual" in generator:
+        compiler_version = settings.get_safe("compiler.version")
+        if compiler_version:
+            compiler_version = compiler_version if "." in compiler_version else \
+                "%s.0" % compiler_version
+            return "Intel C++ Compiler " + compiler_version
```
Comment (ohanar) [inline conans/client/build/cmake_flags.py:33]: This still fails with the following profile
```
[settings]
arch=x86_64
arch_build=x86_64
os=Windows
os_build=Windows
compiler=Visual Studio
compiler.version=16
compiler.runtime=MDd
compiler.toolset=v142
build_type=Debug
[env]
CONAN_CMAKE_GENERATOR=Ninja
```
Granted, it also failed before this PR.

Comment (SSE4) [reply conans/client/build/cmake_flags.py:33]: okay, it was designed to raise on purpose (to detect incompatible config): https://github.com/conan-io/conan/pull/5201
I think I'll leave this specific case as it was before

**Review by danimtb — APPROVED** [review]

**Comment by danimtb** [comment]
@SSE4 please solve conflicts here

**Pull request merged**